homeassistant/components/smappee/config_flow.py

```diff
@@ -21,3 +27,69 @@ class SmappeeFlowHandler(
     def logger(self) -> logging.Logger:
         """Return logger."""
         return logging.getLogger(__name__)
+
+    async def async_step_zeroconf(self, discovery_info=None):
+        """Handle zeroconf discovery."""
+        if discovery_info is None:
+            return self.async_abort(reason="connection_error")
+
+        if not discovery_info[CONF_HOSTNAME].startswith("Smappee1"):
+            # We currently only support Energy and Solar models (legacy)
+            return self.async_abort(reason="invalid mDNS")
+
+        # pylint: E1101: Instance of 'SmappeeFlowHandler' has no 'context' member (no-member)
+        self.context = {"source": "zeroconf"}
+
+        self.context.update(
+            {
+                CONF_HOSTNAME: discovery_info[CONF_HOSTNAME],
+                CONF_TITLE: discovery_info[CONF_HOSTNAME].replace(".local.", ""),
+                CONF_IP_ADDRESS: discovery_info["host"],
+                CONF_SERIALNUMBER: discovery_info[CONF_HOSTNAME]
+                .replace(".local.", "")
+                .replace("Smappee", ""),
+            }
+        )
+
+        # Prepare configuration flow
+        return await self._handle_config_flow(discovery_info, True)
+
+    async def async_step_zeroconf_confirm(self, discovery_info=None):
+        """Handle a flow initiated by zeroconf."""
+        return await self._handle_config_flow(discovery_info)
+
+    async def _handle_config_flow(self, discovery_info=None, prepare=False):
```

Comment:
This helper doesn't seem necessary. I'd move the different parts to the zeroconf step and the confirm step.

Reply:
The config flow still needs other changes. Let's see what remains.